impl/graphsync_test.go

```diff
@@ -1708,6 +1708,126 @@ func TestGraphsyncBlockListeners(t *testing.T) {
+
+	blockChain.VerifyResponseRange(ctx, progressChan, 0, stopPoint)
+	timer := time.NewTimer(100 * time.Millisecond)
+	testutil.AssertDoesReceiveFirst(t, timer.C, "should pause request", progressChan)
+
+	requestID := <-requestIDChan
+
+	// set up extensions and listen for updates on the responder
+	responderExt1 := graphsync.ExtensionData{Name: graphsync.ExtensionName("grip grop"), Data: basicnode.NewString("flim flam, blim blam")}
+	responderExt2 := graphsync.ExtensionData{Name: graphsync.ExtensionName("Humpty/Dumpty"), Data: basicnode.NewInt(101)}
+
+	var responderReceivedExt1 int
+	var responderReceivedExt2 int
+	updateRequests := make(chan struct{}, 1)
+	responder.RegisterRequestUpdatedHook(func(p peer.ID, request graphsync.RequestData, update graphsync.RequestData, hookActions graphsync.RequestUpdatedHookActions) {
+		ext, found := update.Extension(responderExt1.Name)
+		if found {
+			responderReceivedExt1++
+			require.Equal(t, responderExt1.Data, ext)
+		}
+
+		ext, found = update.Extension(responderExt2.Name)
+		if found {
+			responderReceivedExt2++
+			require.Equal(t, responderExt2.Data, ext)
+		}
+
+		updateRequests <- struct{}{}
+	})
+
+	// send updates
+	requestor.SendUpdate(ctx, requestID, responderExt1, responderExt2)
+
+	// check we received what we expected
+	testutil.AssertDoesReceive(ctx, t, updateRequests, "request never completed")
+	require.Equal(t, 1, responderReceivedExt1, "got extension 1 in update")
+	require.Equal(t, 1, responderReceivedExt2, "got extension 2 in update")
+
+	// set up extensions and listen for updates on the requestor
+	requestorExt1 := graphsync.ExtensionData{Name: graphsync.ExtensionName("PING"), Data: basicnode.NewBytes(testutil.RandomBytes(100))}
+	requestorExt2 := graphsync.ExtensionData{Name: graphsync.ExtensionName("PONG"), Data: basicnode.NewBytes(testutil.RandomBytes(100))}
+
+	var requestorReceivedExt1 int
+	var requestorReceivedExt2 int
+
+	requestor.RegisterIncomingResponseHook(func(p peer.ID, responseData graphsync.ResponseData, hookActions graphsync.IncomingResponseHookActions) {
+		ext, found := responseData.Extension(requestorExt1.Name)
+		if found {
+			requestorReceivedExt1++
+			require.Equal(t, requestorExt1.Data, ext)
+		}
+
+		ext, found = responseData.Extension(requestorExt2.Name)
+		if found {
+			requestorReceivedExt2++
+			require.Equal(t, requestorExt2.Data, ext)
+		}
+
+		updateRequests <- struct{}{}
+	})
+
+	// send updates the other way
+	responder.SendUpdate(ctx, requestID, requestorExt1, requestorExt2)
+
+	// check we received what we expected
+	testutil.AssertDoesReceive(ctx, t, updateRequests, "request never completed")
+	require.Equal(t, 1, requestorReceivedExt1, "got extension 1 in update")
+	require.Equal(t, 1, requestorReceivedExt2, "got extension 2 in update")
+
+	// finish up
+	err := responder.Unpause(ctx, requestID)
+	require.NoError(t, err)
+
+	blockChain.VerifyRemainder(ctx, progressChan, stopPoint)
```

Comment:
@hannahhoward I need a bit of help with this test. I'm doing a two-way "update" message, checking that I can send and receive extensions both ways. I've got it working up until here, where the test fails, "response channel never closed". I don't think the unpause is working properly, something about the way I've messed up the response stream perhaps. The logic involved in moving these things around is stretching my comprehension of the internals so there's a good chance I'm not even doing this right at the top level (e.g. is my use of the hooks to receive the extension data the right thing to be doing with these messages?).

Reply:
After I wrote this I had a brainwave that it might have been about my use of hooks + channels, and it turns out that was correct. I was blocking in the test, not inside GS. This passes now.